### PR TITLE
Fix netfx test failures in System.Drawing.Common

### LINKED ISSUE #21294: Test: System.Drawing.Tests.IconTests/ToBitmap_BitmapIcon_Success(icon: (Icon)) failed with "System.ArgumentException"
Opened on behalf of @Jiayili1

The test `System.Drawing.Tests.IconTests/ToBitmap_BitmapIcon_Success(icon: (Icon))` has failed.

System.ArgumentException : Parameter is not valid.

        Stack Trace:

           at System.Drawing.Bitmap..ctor(Int32 width, Int32 height, PixelFormat format)
       at System.Drawing.Icon.BmpFrame()
       at System.Drawing.Icon.ToBitmap()
       at System.Drawing.Tests.IconTests.ToBitmap_BitmapIcon_Success(Icon icon)
Build : Master - 20170621.01 (Core Tests)
Failing configurations:
- Windows.81.Amd64-x86
  - Release

Detail: https://mc.dot.net/#/product/netcore/master/source/official~2Fcorefx~2Fmaster~2F/type/test~2Ffunctional~2Fcli~2F/build/20170621.01/workItem/System.Drawing.Common.Tests/analysis/xunit/System.Drawing.Tests.IconTests~2FToBitmap_BitmapIcon_Success(icon:%20(Icon))

### PR DESCRIPTION
**Pull request by hughbe**
- One case where app switches are flipped between netfx and netcoreapp for PNG icons. I also added a test to make sure netcoreapp also obeys app switches
- One case where an icon would be finalized in an iterator for test data in ToBitmap tests (I think that's what's causing it)
- One case of platform specific serialization

Fixes #21294

### CODE REVIEW

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 1]
```diff
@@ -653,6 +706,7 @@ public void Serialize_RoundtripWithUnownedHandle_Success()
         }
 
         [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "The hardcoded bytes depend on the existance of the System.Drawing.Common assembly.")]
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: typo, existence

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 2]
```diff
@@ -495,40 +492,96 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
+                AppContext.SetSwitch("Switch.System.Drawing.DontSupportPngFramesInIcons", true);
 
-                Assert.Equal(ImageFormat.Png, bitmap.RawFormat);
-                Assert.Equal(77842, bitmap.Flags);
-            }
+                Icon icon;
+                using (var stream = new MemoryStream())
+                {
+                    // Create a PNG inside an ICO.
+                    var bitmap = new Bitmap(10, 10);
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: Bitmaps are also disposable? 
```c#
using (var stream = new MemoryStream())
using (var bitmap = new Bitmap(10, 10);)
...
```

Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: Actually icons are ... gosh, Drawping tests probably should have usings everywhere

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 3]
```diff
@@ -495,40 +492,96 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
+                ToBitmap_BitmapIcon_ReturnsExpected(icon);
+            }
+        }
 
-                // Writing actual data
-                bitmap.Save(stream, ImageFormat.Png);
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        public void ToBitmap_PngIconSupportedInSwitches_Success()
+        {
+            RemoteInvoke(() =>
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: I assume you're going out of proc to isolate other tests from your appcontext switch. That makes sense, but it looks like our other tests don't do that -- even when they switch it both ways...

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 4]
```diff
@@ -495,40 +492,96 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
         }
 
         [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
-        public void ToBitmap_PngIcon_Success()
+        public void ToBitmap_BitmapIconFromHandle_ReturnsExpected()
         {
-            Icon icon;
-            using (var stream = new MemoryStream())
+            // Handle rerring to icon without any colour.
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: typo, referring?

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 3 above)
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: Instead of always going out of proc and checking for isfullframework you could instead check GetSwitch and only go out of proc (and set to false) if it's true.
That's the "right" way to deal with appcontext switches  -- to check the switch, not the version (platform, in this case).

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 5]
```diff
@@ -495,40 +492,96 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
-                // Getting data length (file length minus header)
-                long length = stream.Length - 22;
-                stream.Seek(14, SeekOrigin.Begin);
-                stream.WriteByte((byte)length);
-                stream.WriteByte((byte)(length >> 8));
+                Icon icon;
+                using (var stream = new MemoryStream())
+                {
+                    // Create a PNG inside an ICO.
+                    var bitmap = new Bitmap(10, 10);
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: bitmap is disposable

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 6]
```diff
@@ -495,40 +492,96 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
-                Assert.Empty(bitmap.Palette.Entries);
-                Assert.Equal(icon.Width, bitmap.Width);
-                Assert.Equal(icon.Height, bitmap.Height);
+                AppContext.SetSwitch("Switch.System.Drawing.DontSupportPngFramesInIcons", true);
 
-                Assert.Equal(ImageFormat.Png, bitmap.RawFormat);
-                Assert.Equal(77842, bitmap.Flags);
-            }
+                Icon icon;
+                using (var stream = new MemoryStream())
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: can you avoid duplicating all this code?

**Comment by danmosemsft** [comment]
doh

**Comment by hughbe** [comment]
I've had loads of spare time at work because we're waiting on a change of leadership, so have gone full blast with PRs :D

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: All the handles and stuff are disposed during finalization, but I agree. For now I'll only update usings introduced or changed in this PR though and keep that for another PR

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: I wasn't sure what would happen if the tests ran in parallel so wanted to exclude one change from the other. I suppose this means that the XML tests are racy perhaps.
I've updated the PR to do as suggested

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 6 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: Extracted into a helper

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: Done

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: Fixed

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: Have I always spelled it wrong!

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 7]
```diff
@@ -471,15 +472,11 @@ public static IEnumerable<object[]> ToBitmap_TestData()
             yield return new object[] { new Icon(Helpers.GetTestBitmapPath("256x256_two_entries_multiple_bits.ico"), 48, 48) };
             yield return new object[] { new Icon(Helpers.GetTestBitmapPath("256x256_two_entries_multiple_bits.ico"), 256, 256) };
             yield return new object[] { new Icon(Helpers.GetTestBitmapPath("256x256_two_entries_multiple_bits.ico"), 0, 0) };
```
Comment (stephentoub) [inline src/System.Drawing.Common/tests/IconTests.cs]: Who's Dispose'ing of these Icons?

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 8]
```diff
@@ -495,17 +492,92 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
         [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
-        public void ToBitmap_PngIcon_Success()
+        public void ToBitmap_BitmapIconFromHandle_ReturnsExpected()
+        {
+            // Handle referring to icon without any colour.
+            using (var originalIcon = new Icon(Helpers.GetTestBitmapPath("48x48_one_entry_1bit.ico")))
+            using (Icon icon = Icon.FromHandle(originalIcon.Handle))
+            {
+                ToBitmap_BitmapIcon_ReturnsExpected(icon);
+            }
```
Comment (stephentoub) [inline src/System.Drawing.Common/tests/IconTests.cs]: Why did this one need to be separated out from ToBitmap_TestData?  It'd be worth a comment.

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 9]
```diff
@@ -495,17 +492,92 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
+
+        private const string DontSupportPngFramesInIcons = "Switch.System.Drawing.DontSupportPngFramesInIcons";
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        public void ToBitmap_PngIconSupportedInSwitches_Success()
+        {
+            void VerifyPng()
+            {
+                using (Icon icon = GetPngIcon())
+                {
```
Comment (stephentoub) [inline src/System.Drawing.Common/tests/IconTests.cs]: Nit: unnecessary level of nesting

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 10]
```diff
@@ -495,17 +492,92 @@ public void ToBitmap_BitmapIcon_Success(Icon icon)
...
+                    using (Bitmap bitmap = icon.ToBitmap())
+                    {
+                        Assert.NotSame(icon.ToBitmap(), bitmap);
```
Comment (stephentoub) [inline src/System.Drawing.Common/tests/IconTests.cs]: This Bitmap should be disposed

File: src/System.Drawing.Common/tests/IconTests.cs [hunk 11]
```diff
@@ -653,6 +713,7 @@ public void Serialize_RoundtripWithUnownedHandle_Success()
         }
 
         [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "The hardcoded bytes depend on the existence of the System.Drawing.Common assembly.")]
```
Comment (stephentoub) [inline src/System.Drawing.Common/tests/IconTests.cs]: @morganbr, @mellinoe, is it intentional that these types are serializable?

**Comment by hughbe** [comment]
Addressed PR feedback

**Comment by hughbe** [comment]
@dotnet-bot test this please

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 11 above)
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: No, we should remove [Serializable] from all the drawing types since we don't plan to support it. I thought there was a tracking issue but I can't find it. Of course, with care wrt Mono.

**Comment by danmosemsft** [comment]
@stephentoub any other feedback?

**Comment by danmosemsft** [comment]
Fixes 21294

File: src/System.Drawing.Common/tests/IconTests.cs (hunk 11 above)
Comment (hughbe) [inline src/System.Drawing.Common/tests/IconTests.cs]: Can we do that in another PR?

Comment (danmosemsft) [inline src/System.Drawing.Common/tests/IconTests.cs]: Absolutely 😃

**Comment by danmosemsft** [comment]
@dotnet-bot test Linux x64 Release Build  please (random things several already  tracked)
